sequencer/src/state.rs

```diff
@@ -45,63 +43,102 @@ impl Default for ValidatedState {
+    } = state;
 
-        // fetch receipts from the l1
-        let receipts = fetch_fee_receipts(parent);
-        for FeeReceipt { recipient, amount } in receipts {
-            // Get the balance in order to add amount, ignoring the proof.
-            match fee_merkle_tree.universal_lookup(recipient) {
-                LookupResult::Ok(balance, _) => fee_merkle_tree
-                    .update(recipient, balance.add(amount))
-                    .unwrap(),
-                // Handle `NotFound` and `NotInMemory` by initializing
-                // state.
-                _ => fee_merkle_tree.update(recipient, amount).unwrap(),
-            };
-        }
+    // validate proposal is descendent of parent by appending to parent
+    block_merkle_tree.push(parent.commit()).unwrap();
+    let block_merkle_tree_root = block_merkle_tree.commitment();
+    anyhow::ensure!(
+        proposal.block_merkle_tree_root == block_merkle_tree_root,
+        anyhow::anyhow!(
+            "Invalid Block Root Error: {}, {}",
+            block_merkle_tree_root,
+            proposal.block_merkle_tree_root
+        )
+    );
 
-        let fee_merkle_tree_root = fee_merkle_tree.commitment();
-        anyhow::ensure!(
-            proposal.fee_merkle_tree_root == fee_merkle_tree_root,
-            anyhow::anyhow!(
-                "Invalid Fee Root Error: {}, {}",
-                fee_merkle_tree_root,
-                proposal.fee_merkle_tree_root
-            )
-        );
+    let fee_merkle_tree_root = fee_merkle_tree.commitment();
+    anyhow::ensure!(
+        proposal.fee_merkle_tree_root == fee_merkle_tree_root,
+        anyhow::anyhow!(
+            "Invalid Fee Root Error: {}, {}",
+            fee_merkle_tree_root,
+            proposal.fee_merkle_tree_root
+        )
+    );
+    Ok(())
+}
 
-        Ok(ValidatedState {
-            block_merkle_tree,
-            fee_merkle_tree,
-        })
+/// Fetch receipts from the l1 and add them to local balance.
+fn update_balance(fee_merkle_tree: &mut FeeMerkleTree, parent: &Header) {
+    let receipts = fetch_fee_receipts(parent);
+    for FeeReceipt { recipient, amount } in receipts {
+        // Get the balance in order to add amount, ignoring the proof.
+        match fee_merkle_tree.universal_lookup(recipient) {
+            LookupResult::Ok(balance, _) => fee_merkle_tree
+                .update(recipient, balance.add(amount))
+                .unwrap(),
+            // Handle `NotFound` and `NotInMemory` by initializing
+            // state.
+            _ => fee_merkle_tree.update(recipient, amount).unwrap(),
+        };
     }
 }
+/// Validate builder account by verifiying signature and charging the account.
+fn validate_builder(
+    fee_merkle_tree: &mut FeeMerkleTree,
+    proposed_header: &Header,
+) -> anyhow::Result<()> {
+    let mut verifiable_header = proposed_header.clone();
+    // These unraps should be safe since Header::new() must have
+    // set these fields.
```

Comment:
We cannot trust `Header::new`, it was executed by a different node (the builder, whereas we are acting as a replica here)

Reply:
fixed in 6ea06031645765d2cdcc6542bb4d73afa45f9823